### PR TITLE
Refactor/featurization params

### PR DESCRIPTION
**Pull request by davidegraff**
## Description
This PR updates the featurization module to use dataclasses to store featurization parameters

## Current status
The featurization module currently uses what amounts to a storage container to organize molecule featurization together

## Updated code
The module has now been switched over to using python dataclasses to store these parameters. Specifically, `Featurization_params` has been moved to `FeaturizationParams` and the former `ATOM_FEATURES` `Dict` attribute has been moved to its own dataclass. This was done for a few reasons:
1. **cleanliness:** there's no need to define a length `__init__` function for data classes
2. **inspectability:** dataclasses support automatic `__repr__` definitions. You can now `print(features.PARAMS)` and it will give you this nicely formatted output:
```python
>>> from chemprop import features
>>> print(features.PARAMS)
FeaturizationParams(max_atomic_num=100,
    atom_features=AtomFeaturizationParams(
        atomic_num=[...], degree=[0, 1, 2, 3, 4, 5], formal_charge=[-1, -2, 1, 2, 0],
        chiral_tag=[0, 1, 2, 3], num_Hs=[0, 1, 2, 3, 4], hybridization=[...]
    ),
    path_distance_bins=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], three_d_distance_max=20, three_d_distance_step=1,
    three_d_distance_bins=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20],
    atom_fdim=133, extra_atom_fdim=0, bond_fdim=14,
    extra_bond_fdim=0, reaction_mode=None, explicit_H=False, reaction=False, adding_H=False
)
```
3. **typing:** dataclasses support native type annotation, so we can more easily communicate the value/meaning of each attribute

## Additional updates
This PR also introduces some new changes:
- elimination of a few `get_*` and `set_*` functions in the `featurization` module. All these functions did was mutate the global `PARAMS` variable of the `featurization` module, so for cleanliness, I've removed these. I get that there was probably some intention about hiding/obscuring data with these functions, but it's not the way to do it and just made the code harder to read. I've changed all occurrences to now directly mutate the global `PARAMS` variable
- tied `Featurization_params.max_atomic_num` to the `atomic_num` attribute of the `AtomFeaturizationParams`. The two are logically connected, so the former is now a property that updates the latter whenever it is set. There was some lettuce that needed to be introduced to get this to work, but I believe it's worth the (minimal) clutter
- some unit tests
- renaming of attributes. All the attributes of the `Featurization_params` object were named in all caps before. I think this was to indicate that they were intended to be CONSTANT, but this is python, so they weren't and I've renamed them to be lowercase. However, with the dataclass implementation, we can now actually enforce this, if desired (they've currently been left as proper variables.)

## Checklist
- [x] linted with flake8?
- [x] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by davidegraff** [comment]
looks like I'm failing some numerical checks but I can't pinpoint why

**Comment by davidegraff** [comment]
oddly, I'm not failing *every* integration test, only two of them... Seems to be some numeric instability problems that are resolved when I go back to `master`. The weird thing is that featurization hasn't actually changed- all the default parameters are still the same :/

**Comment by davidegraff** [comment]
hmmm. Any help would be appreciated:
```
$ git checkout master; pytest -k "test_train_single_task_regression_reaction_solvent_2_chemprop_reaction_solvent_explicit_h" -s
Ensemble test rmse = 3.524930
3-fold cross validation
        Seed 0 ==> test rmse = 2.019046
        Seed 1 ==> test rmse = 2.896207
        Seed 2 ==> test rmse = 3.524930
Overall test rmse = 2.813394 +/- 0.617557

$ git checkout refactor/featurization-params; pytest -k "test_train_single_task_regression_reaction_solvent_2_chemprop_reaction_solvent_explicit_h" -s
Ensemble test rmse = 3.520418
3-fold cross validation
        Seed 0 ==> test rmse = 2.298021
        Seed 1 ==> test rmse = 2.876739
        Seed 2 ==> test rmse = 3.520418
Overall test rmse = 2.898393 +/- 0.499277
```
must be something with how this PR affects initial randomness in the library?

**Comment by davidegraff** [comment]
it's the `test_train_single_task_regression_reaction_solvent_2_chemprop_reaction_solvent_explicit_h` and `test_train_single_task_regression_reaction_3_chemprop_reaction_explicit_h` tests that are failing. It doesn't seem like the code breaks, just that the randomness has changed such that we can't make the same assumptions on how the bounds of testing performance. I don't think that's a valid unit/integration test, but the point still stands that this could *should not* affect the random initialization- in principle...

**Comment by cjmcgill** [comment]
Do you see changes in values for all the tests below the 1.5% threshold? Or just reaction ones maybe?

**Comment by davidegraff** [comment]
looks like a few of them i get different results. In some cases my spread is tighter than that of `master`

**Comment by cjmcgill** [comment]
If we can't tighten the reproducibility, it makes me a little uncomfortable about the testing paradigm. I shortened the testing epochs so tests would run faster. Which is fine if we are just recreating a descent curve. But if the descent is unreliable and we actually need a convergence to compare, then we need to test pretty differently. I'm not sure what integration testing looks like in that context.

**Comment by davidegraff** [comment]
Yeah I think these tests are misguided maybe? Logically, this test currently tests for performance bounds. This is a good test to have in the case that the chemprop model provides some theoretical guarantee of performance, but we **don't** have that for the chemprop model. An integration test in this context would mean that the full training/validation/testing pipeline, with all of its complex functions outputs a "reasonable" value. Given that we have no guarantee of what "reasonable" output looks like, we can only test for the fact that it doesn't fail. We might also want to test that our performance isn't abysmal, but there are two possible reasons for terrible performance:
1. operational errors -- these are unit-testable because we can decompose the entire transformation into individual operations that can be unit tested for "correctness". Examples of this could include: improperly featurizing a molecule into a graph, utilizing incorrect elements of a molecular graph for certain network transformations/operations, performing the "wrong" transformation (e.g., not applying a softmax during test time so you're calculating test loss against raw logits for classification)
2. hyperparameter errors -- these are by definition not unit-testable. If I want to specify an incredibly deep and wide network then train for only 1 epoch, is it "wrong" that I get terrible performance?

Anyways- problem still stands that random initialization is somehow changing between the branches and that shouldn't happen

**Comment by cjmcgill** [comment]
It looks like the testing is specifically affected for the tests using the `--explicit_h` argument. I think everything else is consistent, at least down to a 0.1% delta.

**Comment by cjmcgill** [comment]
@davidegraff it looks like something is different with your implementation of `--explicit_h`. I printed out the SMILES associated with the MolGraphs for the first reaction in `test_train_single_task_regression_reaction_solvent_2_chemprop_reaction_solvent_explicit_h`.

For the last previous master:
```
['[C:4](=[C:5]=[O:6])([H:12])[H:13].[O:1]([C:2]([C:3](=[C:7]([H:14])[H:15])[H:11])([H:9])[H:10])[H:8]', '[O:1]([C:2]([C:3]1([H:11])[C:4]([H:12])([H:13])[C:5](=[O:6])[C:7]1([H:14])[H:15])([H:9])[H:10])[H:8]']
```

For your PR:
```
['[CH2:4]=[C:5]=[O:6].[OH:1][CH2:2][CH:3]=[CH2:7]', '[OH:1][CH2:2][CH:3]1[CH2:4][C:5](=[O:6])[CH2:7]1']
```

At the least, it appears to be losing the atom numbering for the hydrogens. So something is implementing different.

**Comment by cjmcgill** [comment]
Sounds good to close for now, but worth keeping this `--explicit_h` thing in back of mind.
